### PR TITLE
Add current working dir as fallback for GitModule

### PR DESCRIPTION
**Pull request by drewnoakes**
(This PR created as [requested in #4809](https://github.com/gitextensions/gitextensions/pull/4809#discussion_r180905799))

When running commands from the command line, and if no working directory argument is passed, use the `Environment.CurrentDirectory` to find a valid git working directory, if possible.

This is important for #4809 where we need values from the config file. Without this change, the editor does cannot read config and therefore does not know git's `core.commentChar` value.

@RussKie pointed out that this change seems to be related to #4740 and asked for it to be pulled out to its own PR. Let's work out what is required here and document it in the code to prevent future flip-flops.

What did I do to test the code and ensure quality:
- Manual testing

Has been tested on:
- GIT 2.16
- Windows 10

### CODE REVIEW

File: GitExtensions/Program.cs [hunk 1]
```diff
@@ -167,6 +168,16 @@ private static string GetWorkingDir(string[] args)
                 }
             }
 
+            if (workingDir == "")
+            {
+                var cwd = Environment.CurrentDirectory;
+                var gitDir = GitModule.FindGitWorkingDir(cwd);
+                if (GitModule.IsValidGitWorkingDir(gitDir))
```
Comment (RussKie) [inline GitExtensions/Program.cs]: not related to the PR, but a general observation - this API is very strange - first we ask fro a working dir, then we double check it is a working dir indeed...

**Comment by drewnoakes** [comment]
Explanation added.

Also simplified code for loading working directory according to your observation.

File: GitCommands/Git/GitModule.cs [hunk 2]
```diff
@@ -479,17 +479,17 @@ List<string> GetSubmodulePaths(GitModule module)
...
 
-        [NotNull]
-        public static string FindGitWorkingDir([CanBeNull] string startDir)
+        /// <summary>
+        /// Searches from <paramref name="startDir"/> and up through the directory
+        /// hierarchy for a valid git working directory. If found, the path is returned,
+        /// otherwise <c>null</c>.
+        /// </summary>
+        [CanBeNull]
+        public static string TryFindGitWorkingDir([CanBeNull] string startDir)
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs]: Can we write some sort of unit test for this?

File: GitExtensions/Program.cs [hunk 3]
```diff
@@ -159,14 +160,22 @@ private static string GetWorkingDir(string[] args)
...
-            if (args.Length <= 1 && string.IsNullOrEmpty(workingDir) && AppSettings.StartWithRecentWorkingDir)
+            if (args.Length <= 1 && workingDir == null && AppSettings.StartWithRecentWorkingDir)
             {
                 if (GitModule.IsValidGitWorkingDir(AppSettings.RecentWorkingDir))
                 {
                     workingDir = AppSettings.RecentWorkingDir;
                 }
             }
 
+            if (workingDir == null)
```
Comment (gerhardol) [inline GitExtensions/Program.cs]: Tested too: This breaks AppSettings.StartWithRecentWorkingDir (and reverses #4740 as well as "fixes" 4860).
But the AppSettings.StartWithRecentWorkingDir can maybe be added somewhere else.
(Looking at it, rebasing on to of master.)
Maybe the PR should be merged and fixed later?

The following should be changed too:

            if (args.Length <= 1 && workingDir == null && AppSettings.StartWithRecentWorkingDir)
....
            if (workingDir == null)

**Comment by gerhardol** [comment]
I pushed an update to my branch. As those changes mostly applies to the first commit, I rebased/squashed my changes to https://github.com/gerhardol/gitextensions/tree/drewnoakes/_working-dir-cwd-rebase

I could create a new PR too.

**Comment by RussKie** [comment]
@gerhardol please do. I think @drewnoakes is currently MIA busy with his day job.

**Comment by gerhardol** [comment]
Thanks @drewnoakes 
Looking forward to get you back

**Pull request merged**